hudi-spark/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/HoodieParquetRecordReaderIterator.scala

```diff
@@ -0,0 +1,178 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.parquet
+
+import org.apache.hadoop.mapred.JobConf
+import org.apache.hudi.common.fs.FSUtils
+import org.apache.hudi.common.table.log.{HoodieMergedLogRecordScanner, LogReaderUtils}
+import org.apache.hudi.hadoop.config.HoodieRealtimeConfig
+import org.apache.hudi.hadoop.realtime.HoodieRealtimeFileSplit
+import org.apache.parquet.hadoop.ParquetRecordReader
+
+import org.apache.avro.Schema
+import org.apache.hudi.hadoop.utils.HoodieInputFormatUtils.HOODIE_RECORD_KEY_COL_POS
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.avro.{AvroDeserializer, SchemaConverters}
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.{UnsafeProjection, UnsafeRow}
+import org.apache.spark.sql.types.StructType
+import java.io.Closeable
+import java.util
+
+import scala.util.Try
+
+/**
+ * This class is the iterator for Hudi MOR table.
+ * Log files are scanned on initialization.
+ * This iterator will read the parquet file first and skip the record if it present in the log file.
+ * Then read the log file.
+ * Custom payload is not supported yet. This combining logic is matching with [OverwriteWithLatestAvroPayload]
+ * @param rowReader HoodieRealtimeParquetRecordReader
+ */
+class HoodieParquetRecordReaderIterator(private[this] var rowReader: ParquetRecordReader[UnsafeRow],
+                                        private[this] val split: HoodieRealtimeFileSplit,
+                                        private[this] val jobConf: JobConf) extends Iterator[UnsafeRow] with Closeable with Logging {
+  private[this] var havePair = false
+  private[this] var finished = false
+  private[this] var parquetFinished = false
+
+  private[this] var deltaRecordMap: util.Map[String,org.apache.hudi.common.model
+  .HoodieRecord[_ <: org.apache.hudi.common.model.HoodieRecordPayload[_ <: org.apache.hudi.common.model.HoodieRecordPayload[_ <: AnyRef]]]] = _
+  private[this] var deltaRecordKeys: util.Set[String] = _
+  private[this] var deltaIter: util.Iterator[String] = _
+  private[this] var avroSchema: Schema = _
+  private[this] var sparkTypes: StructType = _
+  private[this] var converter: AvroDeserializer = _
+
+  // SPARK-23457 Register a task completion lister before `initialization`.
```

Comment:
does this comment make sense in this context?

Reply:
yea, doesn't really make sense to put it here